File: frontend/src/api/portfolio.ts
// Client API pour le portefeuille (`GET /api/portfolio/*`, B18 backend —
// volontairement resté backend-only à l'époque, voir docstring de
// `backend/app/routers/portfolio.py` : "B26 consommera ces routes"). Ce
// module frontend n'existait pas avant B26 : premier consommateur réel de
// ces routes.

import { apiGet } from "./client";

export type PortfolioSummary = {
  cash: number;
  buying_power: number;
  portfolio_value: number;
  // `null` tant que le worker n'a pas encore assez de tours — jamais
  // fabriqué en 0 (voir backend/app/schemas/portfolio.py).
  daily_pl: number | null;
  total_pl: number | null;
  snapshot_at: string;
};

export type Position = {
  symbol: string;
  quantity: number;
  average_entry_price: number | null;
  market_value: number | null;
  unrealized_pl: number | null;
  snapshot_at: string;
};

export type PositionsResponse = {
  positions: Position[];
  snapshot_at: string | null;
};

export type PerformanceCard = {
  window: "1D" | "3D" | "7D";
  available: boolean;
  reason?: string | null;
  value_change?: number | null;
  percent_change?: number | null;
};

export type PerformanceCardsResponse = {
  cards: PerformanceCard[];
};

export type PortfolioHistoryItem = {
  cash: number;
  buying_power: number;
  portfolio_value: number;
  daily_pl: number | null;
  total_pl: number | null;
  snapshot_at: string;
};

export type PortfolioHistoryResponse = {
  items: PortfolioHistoryItem[];
  total: number;
  limit: number;
  offset: number;
};

export async function fetchPortfolioSummary(): Promise<PortfolioSummary> {
  return apiGet<PortfolioSummary>("/api/portfolio/summary");
}

export async function fetchPositions(): Promise<PositionsResponse> {
  return apiGet<PositionsResponse>("/api/portfolio/positions");
}

export async function fetchPerformanceCards(): Promise<PerformanceCardsResponse> {
  return apiGet<PerformanceCardsResponse>("/api/portfolio/performance");
}

// §B27 "Courbe portefeuille"/"Sparklines 1D-7D" — premier consommateur
// frontend de `GET /api/portfolio/history` (B18, backend-only jusqu'ici).
export async function fetchPortfolioHistory(limit = 200): Promise<PortfolioHistoryResponse> {
  return apiGet<PortfolioHistoryResponse>(`/api/portfolio/history?limit=${limit}`);
}

File: backend/app/schemas/portfolio.py
"""Schémas Pydantic — B18 (Portefeuille, positions et historique)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class PortfolioSummaryOut(BaseModel):
    cash: float
    buying_power: float
    portfolio_value: float
    # §B18 — honnêtement `None` tant que le worker n'a pas encore au moins
    # deux tours pour `daily_pl` (voir `_compute_daily_pl`), ou aucun
    # snapshot antérieur pour `total_pl` — jamais fabriqués.
    daily_pl: float | None
    total_pl: float | None
    snapshot_at: datetime

    model_config = {"from_attributes": True}


class PositionOut(BaseModel):
    symbol: str
    quantity: float
    average_entry_price: float | None
    market_value: float | None
    unrealized_pl: float | None
    snapshot_at: datetime

    model_config = {"from_attributes": True}


class PositionsResponse(BaseModel):
    positions: list[PositionOut]
    # `None` si aucun tour de positions n'a encore été écrit pour ce
    # contexte — distinct de "compte sans aucune position ouverte" (auquel
    # cas `snapshot_at` est renseigné mais `positions` est vide).
    snapshot_at: datetime | None


class PortfolioHistoryItemOut(BaseModel):
    cash: float
    buying_power: float
    portfolio_value: float
    daily_pl: float | None
    total_pl: float | None
    snapshot_at: datetime

    model_config = {"from_attributes": True}


class PortfolioHistoryResponse(BaseModel):
    items: list[PortfolioHistoryItemOut]
    total: int
    limit: int
    offset: int


class PerformanceCardOut(BaseModel):
    # "1D", "3D", "7D" (§B18 "Cartes 1D à 7D").
    window: str
    available: bool
    # Toujours "Not enough account history yet" quand `available=False` —
    # texte exact du libellé §18 de la spec, jamais une variation fabriquée.
    reason: str | None = None
    value_change: float | None = None
    percent_change: float | None = None


class PerformanceCardsResponse(BaseModel):
    cards: list[PerformanceCardOut]
